src/DataSource.ts

```diff
@@ -40,39 +39,38 @@ export class DataSource extends DataSourceApi<MyQuery> {
 
   async query(options: DataQueryRequest<MyQuery>): Promise<DataQueryResponse> {
     const { range } = options;
-    const from = range!.from.unix();
-    const to = range!.to.unix();
-    const datasource = this; // defined to be reachable on the next closure
+    const from = range.from.unix();
+    const to = range.to.unix();
 
     const promises = options.targets.map((target) => {
       const query = defaults(target, defaultQuery);
-      return datasource.getGraphQuery([from, to], query);
+      return this.getGraphQuery([from, to], query);
     });
     return Promise.all(promises).then((data) => ({ data }));
   }
 
-  async getGraphQuery(range: number[], query: MyQuery) {
+  async getGraphQuery(range: number[], query: MyQuery): Promise<MutableDataFrame<unknown>> {
     if (isEmpty(query.context) || !query.params.graph_name) {
       return new MutableDataFrame();
     }
     const editionMode = get(this, 'instanceSettings.jsonData.edition', 'CEE');
-    const response = await this.doRequest({
+    const response = await this.doRequest<ResponseDataCurves>({
       ...query,
       params: { action: 'get_graph' },
       data: graphDefinitionRequest(editionMode, query, range),
     });
     return buildMetricDataFrame(response, query);
   }
 
-  async testDatasource() {
+  async testDatasource(): Promise<unknown | undefined> {
```

Comment:
Instead of `unkown`, I'd use the type of the return object in the `then` part.

Reply:
Hm. The base class defines this as `Promise<any>` so it seems okayish to return anything, so I thought it would be good to not type this explicitly.

Reply:
Given that the Type is under specified in the API, your change is good.